Sources/ServiceLifecycle/ServiceGroup.swift

```diff
@@ -200,73 +235,174 @@ public actor ServiceGroup: Sendable {
+                            )
+                            group.cancelAll()
+                            return .success(())
+                        }
+                    }
 
-                    do {
-                        try await self.shutdownGracefully(
-                            group: &group,
-                            gracefulShutdownManagers: gracefulShutdownManagers
+                case .serviceThrew(let service, let index, let error):
+                    switch service.throwBehaviour.behavior {
+                    case .cancelGroup:
+                        self.logger.error(
+                            "Service threw error. Cancelling group.",
+                            metadata: [
+                                self.loggingConfiguration.keys.serviceKey: "\(service.service)",
+                                self.loggingConfiguration.keys.errorKey: "\(error)",
+                            ]
                         )
-                    } catch {
+                        group.cancelAll()
                         return .failure(error)
+
+                    case .gracefullyShutdownGroup:
+                        self.logger.error(
+                            "Service threw error. Shutting down group.",
+                            metadata: [
+                                self.loggingConfiguration.keys.serviceKey: "\(service.service)",
+                                self.loggingConfiguration.keys.errorKey: "\(error)",
+                            ]
+                        )
+                        services[index] = nil
+
+                        do {
+                            try await self.shutdownGracefully(
+                                services: services,
+                                group: &group,
+                                gracefulShutdownManagers: gracefulShutdownManagers
+                            )
+                        } catch {
+                            return .failure(error)
+                        }
+
+
+                    case .ignore:
+                        self.logger.debug(
+                            "Service threw error.",
+                            metadata: [
+                                self.loggingConfiguration.keys.serviceKey: "\(service.service)",
+                                self.loggingConfiguration.keys.errorKey: "\(error)",
+                            ]
+                        )
+                        services[index] = nil
+
+                        if services.allSatisfy({ $0 == nil }) {
+                            self.logger.debug(
+                                "All services finished."
+                            )
+
+                            group.cancelAll()
+                            return .success(())
+                        }
+                    }
+
+                case .signalCaught(let unixSignal):
+                    if self.gracefulShutdownSignals.contains(where: { $0 == unixSignal }) {
```

Comment:
Isn't it possible to just check `self.gracefulShutdownSignals.contains(unixSignal)`?